Feature imgui builder

StuartSulaski opened the pull request:
Cleaner integration of Dear Imgui with builders.
Can be used by the user to generate debug UI

Quentin01 committed:
Add method Core::Version::fromInt(uint32_t)

StuartSulaski committed:
Updated CMakeList to unclude imgui header and source

StuartSulaski committed:
Updated thirdparty submodule and fixed .gitmodules

StuartSulaski committed:
updated third party reference

StuartSulaski committed:
Gui::init() done and CMakeList updated

Updated the CMakeList to corectly include the imgui source
Rewrote Gui::init()

StuartSulaski committed:
Started redoing initFontsTexture()

StuartSulaski committed:
Finished rewritting initFontsTexture

StuartSulaski committed:
Refactor gui using builders and API::CommandBuffer methods

Nokitoo committed:
Clean Gui code

StuartSulaski committed:
Updated Macro.cmake so that sample include imgui headers and source

StuartSulaski committed:
Removing unnecessary lines in Macros.cmake and updating thirdparty

StuartSulaski committed:
updated thirdparty submodule

StuartSulaski committed:
Fixed issue that was causing a build error

StuartSulaski committed:
Pushing gui shaders

Quentin01 commented on include/lug/Core/Version.inl:
```diff
@@ -4,4 +4,5 @@ inline Version Version::fromInt(uint32_t version) {
         (version >> 12) & 0x3FF, // minor
         version & 0xFFF // patch
     };
+
```
Why did you add a line?

Quentin01 commented on include/lug/Core/Application.hpp:
```diff
@@ -158,7 +158,7 @@ class LUG_CORE_API Application {
 
 private:
-    bool beginFrame();
+    bool beginFrame(const lug::System::Time &elapsedTime);
```
No space between the type and `&`

Quentin01 commented on include/lug/Graphics/Render/Window.hpp:
```diff
@@ -28,7 +29,7 @@ class LUG_GRAPHICS_API Window: public ::lug::Window::Window, public ::lug::Graph
     virtual ~Window() = default;
 
-    virtual bool beginFrame() = 0;
+    virtual bool beginFrame(const lug::System::Time &elapsedTime) = 0;
```
No space between the type and `&`

Quentin01 commented on include/lug/Graphics/Renderer.hpp:
```diff
@@ -37,7 +37,7 @@ class LUG_GRAPHICS_API Renderer {
     virtual bool finishInit() = 0;
 
-    virtual bool beginFrame() = 0;
+    virtual bool beginFrame(const lug::System::Time &elapsedTime) = 0;
```
No space between the type and `&`

Quentin01 commented on include/lug/Graphics/Vulkan/API/Builder/GraphicsPipeline.inl:
```diff
@@ -213,9 +213,12 @@ inline void GraphicsPipeline::setShader(VkShaderStageFlagBits stage, const char*
-    shaderModuleBuilder.loadFromFile(filename);
+    VkResult result{ VK_SUCCESS };
+    if (!shaderModuleBuilder.loadFromFile(filename)) {
+        LUG_LOG.error("Builder::GraphicsPipeline: Create load from file {}: {}", filename, result);
```
You don't need to use `result` in this error message because it will always be `VK_SUCCESS`.
So, keep the initialization of `result` as before and just remove it from this message.

Quentin01 commented on include/lug/Graphics/Vulkan/API/Builder/Sampler.hpp:
```diff
@@ -56,7 +56,7 @@ class Sampler {
     float _mipLodBias{0.0f};
     VkBool32 _anisotropyEnable{VK_FALSE};
-    float _maxAnisotropy{0.0f};
+    float _maxAnisotropy{1.0f};
```
Why did you change the default value? Just use `setMaxAnisotropy` in your code.

StuartSulaski replied:
Vulkan wishes the default value to be set at 1.0f

Quentin01 commented on include/lug/Graphics/Vulkan/Render/Window.hpp:
```diff
@@ -96,6 +98,8 @@ class LUG_GRAPHICS_API Window final : public ::lug::Graphics::Render::Window {
 
     API::CommandPool _commandPool{};
+
+    std::unique_ptr<lug::Graphics::Vulkan::Gui>   _guiInstance;
```
Is it really mandatory to use a pointer and not just an instance?
And one useless space.

StuartSulaski replied:
Yes because I need to forward declare it

Quentin01 commented on include/lug/Graphics/Vulkan/Gui.hpp:
```diff
@@ -0,0 +1,96 @@
+
+    Gui(Renderer& renderer, Render::Window &window);
+
+    Gui(const Gui&) = delete;
```
Where is the constructor by movement?

Quentin01 commented on include/lug/Graphics/Vulkan/Gui.hpp:
```diff
@@ -0,0 +1,96 @@
+    };
+
+public:
+    Gui() = delete;
```
You don't need to specify this one if you specify another constructor

Quentin01 commented on include/lug/Graphics/Vulkan/Gui.hpp:
```diff
@@ -0,0 +1,96 @@
+    bool beginFrame(const lug::System::Time& elapsedTime);
+    bool endFrame(const std::vector<VkSemaphore>& waitSemaphores, uint32_t currentImageIndex);
+
+    void processEvents(lug::Window::Event event);
```
`processEvent` because you are just processing one event
And use a `const lug::Window::Event&` instead.

Quentin01 commented on include/lug/Graphics/Vulkan/Gui.hpp:
```diff
@@ -0,0 +1,96 @@
+
+    void processEvents(lug::Window::Event event);
+
+    const Vulkan::API::Semaphore& getGuiSemaphore(uint32_t currentImageIndex) const;
```
Is it really useful to repeat `Gui` in the name of this method?
Maybe rename it `getEndDrawSemaphore` or something like that.

Quentin01 commented on include/lug/Graphics/Vulkan/Gui.hpp:
```diff
@@ -0,0 +1,96 @@
+    Render::Window& _window;
+
+    const API::Queue* _presentQueue{nullptr};
+    const API::QueueFamily* _presentQueueFamily{ nullptr};
```
No space between `{` and the value.

Quentin01 commented on include/lug/Graphics/Vulkan/Gui.hpp:
```diff
@@ -0,0 +1,96 @@
+    API::ImageView _fontImageView;
+    API::DeviceMemory _fontDeviceMemory;
+    API::Sampler _fontSampler;
+
```
Why do you have two newlines instead of one?

Quentin01 commented on include/lug/Graphics/Vulkan/Render/Window.hpp:
```diff
@@ -51,7 +52,7 @@ class LUG_GRAPHICS_API Window final : public ::lug::Graphics::Render::Window {
     bool pollEvent(lug::Window::Event&) override;
 
-    bool beginFrame() override final;
+    bool beginFrame(const lug::System::Time &elapsedTime) override final;
```
No space between the type and `&`

Quentin01 commented on include/lug/Graphics/Vulkan/Renderer.hpp:
```diff
@@ -86,7 +85,7 @@ class LUG_GRAPHICS_API Renderer final : public ::lug::Graphics::Renderer {
     void destroy();
 
-    bool beginFrame() override final;
+    bool beginFrame(const lug::System::Time &elapsedTime) override final;
```
No space between the type and `&`

Quentin01 commented on resources/shaders/gui.vert:
```diff
@@ -0,0 +1,25 @@
+layout (location = 0) out vec2 outUV;
+layout (location = 1) out vec4 outColor;
+
+out gl_PerVertex
```
In Vulkan you don't need to specify that.

Quentin01 commented on samples/Macros.cmake:
```diff
@@ -124,6 +124,16 @@ macro(lug_add_sample target)
 
+    # find imgui
+    if (NOT EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/../../thirdparty/imgui")
+    message(FATAL_ERROR "Can't find imgui, call `git submodule update --recursive`")
```
Indentation

Quentin01 commented on samples/hello/src/Application.cpp:
```diff
@@ -7,6 +7,8 @@
 #include <lug/Graphics/Renderer.hpp>
 #include <lug/Graphics/Vulkan/Renderer.hpp>
 
+#include <imgui.h>
```
Order of inclusion

Quentin01 commented on samples/hello/src/Application.cpp:
```diff
@@ -169,4 +171,6 @@ void Application::onFrame(const lug::System::Time& elapsedTime) {
         renderViews[i]->getCamera()->lookAt({0.0f, 0.0f, 0.0f}, {0.0f, 1.0f, 0.0f}, lug::Graphics::Node::TransformSpace::World);
     }
+
+    ImGui::ShowTestWindow();
```
Do we really want this test window in our sample? If yes, it's probably better to put it in another sample.

Quentin01 committed:
Remove useless files

Quentin01 commented on src/lug/Core/Application.cpp:
```diff
@@ -77,8 +77,8 @@ void Application::close() {
 
-bool Application::beginFrame() {
-    return _graphics.getRenderer()->beginFrame();
+bool Application::beginFrame(const lug::System::Time &elapsedTime) {
```
No space between the type and `&`

Quentin01 commented on src/lug/Graphics/CMakeLists.txt:
```diff
@@ -90,12 +90,12 @@ set(SRC
     ${SRCROOT}/Vulkan/Requirements/Core.hpp
     ${SRCROOT}/Vulkan/Requirements/Requirements.hpp
 
+    ${SRCROOT}/Vulkan/Renderer.cpp
```
Why did you change its position? It's ordered by alphabetic order

Quentin01 commented on src/lug/Graphics/CMakeLists.txt:
```diff
@@ -90,12 +90,12 @@ set(SRC
 
+    ${SRCROOT}/Vulkan/Renderer.cpp
     ${SRCROOT}/Vulkan/Vulkan.cpp
+    ${SRCROOT}/Vulkan/Gui.cpp
```
Alphabetic order

Quentin01 commented on src/lug/Graphics/CMakeLists.txt:
```diff
@@ -243,6 +243,7 @@ set(INC
     ${INCROOT}/Vulkan/Renderer.hpp
     ${INCROOT}/Vulkan/Renderer.inl
     ${INCROOT}/Vulkan/Vulkan.hpp
+    ${INCROOT}/Vulkan/Gui.hpp
```
Alphabetic order

Quentin01 commented on src/lug/Graphics/Vulkan/API/Device.cpp:
```diff
@@ -67,6 +67,14 @@ API::QueueFamily* Device::getQueueFamily(VkQueueFlags flags, bool supportPresent
+const API::Queue* Device::getQueue(const std::string & queueName) const {
+    for (const API::QueueFamily& queueFamily : _queueFamilies) {
+        const API::Queue* queue = queueFamily.getQueue(queueName);
+        if (queue) return queue;
```
Even if it's a one line `if`, you need to use the brackets.

Quentin01 commented on src/lug/Graphics/Vulkan/API/Device.cpp:
```diff
@@ -67,6 +67,14 @@ API::QueueFamily* Device::getQueueFamily(VkQueueFlags flags, bool supportPresent
+        const API::Queue* queue = queueFamily.getQueue(queueName);
+        if (queue) return queue;
+    }
+    return nullptr;
```
Put a newline between the bracket and the `return`.

Quentin01 commented on src/lug/Graphics/Vulkan/API/Queue.cpp:
```diff
@@ -69,6 +73,11 @@ bool Queue::waitIdle() const {
 }
 
+const QueueFamily * Queue::getQueueFamily() const
+{
```
Bracket at the end of the previous line

Quentin01 commented on src/lug/Graphics/Vulkan/Render/Window.cpp:
```diff
@@ -10,6 +10,7 @@
 #include <lug/Graphics/Vulkan/API/Builder/Swapchain.hpp>
 #include <lug/Graphics/Vulkan/API/Instance.hpp>
 #include <lug/System/Logger/Logger.hpp>
+#include <lug/Graphics/Vulkan/Gui.hpp>
```
Order of inclusion

Quentin01 commented on src/lug/Graphics/Vulkan/Render/Window.cpp:
```diff
@@ -41,13 +42,16 @@ bool Window::pollEvent(lug::Window::Event& event) {
 
-bool Window::beginFrame() {
+bool Window::beginFrame(const lug::System::Time &elapsedTime) {
+    _guiInstance->beginFrame(elapsedTime);
```
It's probably better to do that at the end of the function (during the function we kind of refresh the swapchain if it is outdated), and you also need to check the return of the function because it returns a `bool`.

Quentin01 commented on src/lug/Graphics/Vulkan/Render/Window.cpp:
```diff
@@ -64,7 +68,7 @@ bool Window::beginFrame() {
                 return false;
             }
-
+            _guiInstance->initFramebuffers(_swapchain.getImagesViews());
```
Check the return value

Quentin01 commented on src/lug/Graphics/Vulkan/Render/Window.cpp:
```diff
@@ -286,6 +292,17 @@ bool Window::initSwapchainCapabilities() {
 }
 
+bool Window::initGui()
+{
```
Bracket at the end of the previous line

Quentin01 commented on src/lug/Graphics/Vulkan/Render/Window.cpp:
```diff
@@ -286,6 +292,17 @@ bool Window::initSwapchainCapabilities() {
+{
+    _guiInstance = std::make_unique<Gui>(_renderer, *this);
+    if (!_guiInstance->init(_swapchain.getImagesViews()))
+    {
```
Bracket at the end of the previous line

Quentin01 commented on src/lug/Graphics/Vulkan/Render/Window.cpp:
```diff
@@ -286,6 +292,17 @@ bool Window::initSwapchainCapabilities() {
+        LUG_LOG.error("RendererWindow: Can't create font texture for Gui");
+        return false;
+    }
+    return true;
```
Newline between the bracket and the `return`.

Quentin01 commented on src/lug/Graphics/Vulkan/Render/Window.cpp:
```diff
@@ -286,6 +292,17 @@ bool Window::initSwapchainCapabilities() {
+    _guiInstance = std::make_unique<Gui>(_renderer, *this);
+    if (!_guiInstance->init(_swapchain.getImagesViews()))
+    {
+        LUG_LOG.error("RendererWindow: Can't create font texture for Gui");
```
The message isn't good

Quentin01 commented on src/lug/Graphics/Vulkan/Render/Window.cpp:
```diff
@@ -580,6 +597,12 @@ bool Window::initRender() {
     }
 
+    if (!initGui())
+    {
```
Bracket at the end of the previous line.

Quentin01 commented on src/lug/Graphics/Vulkan/Render/Window.cpp:
```diff
@@ -580,6 +597,12 @@ bool Window::initRender() {
         }
     }
 
+    if (!initGui())
```
Why don't you call this function with all the others?

Quentin01 commented on src/lug/Graphics/Vulkan/Renderer.cpp:
```diff
@@ -565,8 +565,8 @@ ::lug::Graphics::Render::Window* Renderer::getWindow() {
 
-bool Renderer::beginFrame() {
-    return _window->beginFrame();
+bool Renderer::beginFrame(const lug::System::Time &elapsedTime) {
```
No space between the type and `&`

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+namespace Vulkan {
+
+Gui::Gui(Renderer & renderer, Render::Window & window) : _renderer(renderer), _window(window) {
+}
```
Bracket at the end of the previous line

StuartSulaski replied:
Isn't this already correct ?

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+namespace Graphics {
+namespace Vulkan {
+
+Gui::Gui(Renderer & renderer, Render::Window & window) : _renderer(renderer), _window(window) {
```
No space between the type and `&`

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+            return false;
+        }
+    }
+    API::Builder::CommandBuffer commandBufferBuilder(_renderer.getDevice(), _commandPool);
```
Newline between the bracket and this line.

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+    }
+    API::Builder::CommandBuffer commandBufferBuilder(_renderer.getDevice(), _commandPool);
+    commandBufferBuilder.setLevel(VK_COMMAND_BUFFER_LEVEL_PRIMARY);
+
```
Remove one empty line

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+    for (uint32_t i = 0; i < imageViews.size(); ++i) {
+        _framesData[i].commandBuffer = std::move(commandBuffers[i]);
+    }
+    commandBuffers.resize(0); // clear the array
```
Why not using `clear` instead?

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+    ImGuiIO& io = ImGui::GetIO();
+
+    // Create font texture
+    unsigned char* fontData;
```
Even if it's reset later, always initialize a variable with a value.

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+
+    // Create font texture
+    unsigned char* fontData;
+    int texWidth, texHeight;
```
Even if it's reset later, always initialize a variable with a value.

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+    unsigned char* fontData;
+    int texWidth, texHeight;
+    io.Fonts->GetTexDataAsRGBA32(&fontData, &texWidth, &texHeight);
+    size_t uploadSize = texWidth * texHeight * 4 * sizeof(char);
```
You can use a `const size_t`

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+
+            result = VK_SUCCESS;
+            if (!deviceMemoryBuilder.build(_fontDeviceMemory, &result)) {
+                LUG_LOG.error("Gui::initFontsTexture: Can't create staging buffer device memory: {}", result);
```
The error message is wrong

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+        // Create staging buffer
+        {
+            API::Builder::Buffer bufferBuilder(device);
+            bufferBuilder.setQueueFamilyIndices(queueFamilyIndices);
```
Why do you use an external variable?

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+                return false;
+            }
+        }
+
```
Remove one of the empty line

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+    io.DisplaySize = ImVec2(_window.getWidth(), _window.getHeight());
+    io.DisplayFramebufferScale = ImVec2(1.0f, 1.0f);
+
+    // Get present queue family and retrieve the first queue
```
It's just a graphic queue that you want, not the present queue. So rename the variables, etc to match that.

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+        imageBuilder.setUsage(VK_IMAGE_USAGE_SAMPLED_BIT | VK_IMAGE_USAGE_TRANSFER_DST_BIT);
+        imageBuilder.setPreferedFormats({ VK_FORMAT_R8G8B8A8_UNORM, VK_FORMAT_D32_SFLOAT, VK_FORMAT_D32_SFLOAT_S8_UINT, VK_FORMAT_D24_UNORM_S8_UINT });
+        imageBuilder.setFeatureFlags(VK_FORMAT_FEATURE_SAMPLED_IMAGE_BIT);
+        imageBuilder.setTiling(VK_IMAGE_TILING_OPTIMAL);
```
You need to use `setQueueFamilyIndices` also with the id of the transfer queue and the id of the graphic queue

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+            VkResult result{VK_SUCCESS};
+            API::CommandBuffer commandBuffer;
+
+            API::Builder::CommandBuffer commandBufferBuilder(_renderer.getDevice(), _commandPool);
```
You can't use `_commandPool` if the transfer queue family and the graphic queue family are not the same

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+            bufferCopyRegion.imageExtent.height = texHeight;
+            bufferCopyRegion.imageExtent.depth = 1;
+
+            vkCmdCopyBufferToImage(
```
Add this function to the class `API::CommandBuffer`

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+            }
+
+            // Copy
+            VkBufferImageCopy bufferCopyRegion = {};
```
To initialize a structure, use the same as in the guideline.

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+        API::Builder::DeviceMemory deviceMemoryBuilder(device);
+        deviceMemoryBuilder.setMemoryFlags(VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT);
+
+        VkExtent3D extent{
```
To initialize a structure, use the same as in the guideline. (You can also use a const)

```cpp
const VkExtent3D extent{
    /* extent.width */ texWidth,
    /* extent.height */ texHeight,
    /* extent.depth */ 1
};
```

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+                commandBuffer.pipelineBarrier(pipelineBarrier);
+            }
+
+            commandBuffer.end();
```
Test if `end` fails

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+        {
+            API::Builder::Semaphore semaphoreBuilder(device);
+
+            VkResult result;
```
Initialize the result to `VK_SUCCESS`

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+            API::Builder::Fence fenceBuilder(device);
+            fenceBuilder.setFlags(VK_FENCE_CREATE_SIGNALED_BIT); // Signaled state
+
+            VkResult result;
```
Initialize the result to `VK_SUCCESS`

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+        }
+    }
+
+    for (auto& frameData: _framesData) {
```
Why do you initialize the framedatas in a function named `initFontsTexture`?

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+
+        _descriptorSet.updateImages(0, 0, VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER, {descriptorImageInfo});
+
+        VkPushConstantRange pushConstant = {
```
See the guideline to initialize a structure

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+
+    io.DisplaySize = ImVec2(_window.getWidth(), _window.getHeight());
+
+    uint32_t x;
```
Always initialize a variable

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+    io.DisplaySize = ImVec2(_window.getWidth(), _window.getHeight());
+
+    uint32_t x;
+    uint32_t y;
```
Always initialize a variable

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+        return false;
+    }
+    frameData.fence.reset();
+    updateBuffers(currentImageIndex);
```
Check the return value (this function need to return a `bool`)

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+
+    API::Device &device = _renderer.getDevice();
+
+    API::QueueFamily* graphicsQueueFamily;
```
You won't need this anymore when you will rename `_presentQueueFamily` in `_graphicsQueueFamily`

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+    {
+        graphicsQueueFamily = device.getQueueFamily(VK_QUEUE_GRAPHICS_BIT);
+        if (!graphicsQueueFamily) {
+            LUG_LOG.error("Gui::initFontsTexture: Can't find VK_QUEUE_GRAPHICS_BIT queue family");
```
It's not `initFontsTexture`

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+        }
+        graphicsQueue = &graphicsQueueFamily->getQueues().front();
+        if (!graphicsQueue) {
+            LUG_LOG.error("Gui::initFontsTexture: Can't find graphics queue");
```
It's not `initFontsTexture`

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+    frameData.commandBuffer.bindIndexBuffer(frameData.indexBuffer, VK_INDEX_TYPE_UINT16);
+
+    // UI scale and translate via push constants
+    PushConstBlock pushConstants{
```
You can use a const

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+void Gui::processEvents(lug::Window::Event event) {
+    ImGuiIO& io = ImGui::GetIO();
+    switch (event.type) {
+    case (lug::Window::Event::Type::KeyPressed):
```
Indentation, and do not use parentheses around `case` value.

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+        io.KeySuper = static_cast<int>(event.key.system);
+        break;
+    case (lug::Window::Event::Type::CharEntered):
+        if (event.character.val > 0 && event.character.val < 0x10000) {
```
Why do you have this condition? Can't you use `isprint`?

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+    return _framesData[currentImageIndex].semaphore;
+}
+
+void Gui::updateBuffers(uint32_t currentImageIndex) {
```
This function needs to return a `bool`

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+    // Update buffers only if vertex or index count has been changed compared to current buffer size
+    {
+        // Vertex buffer
+        if ((static_cast<VkBuffer>(frameData.vertexBuffer) == VK_NULL_HANDLE) || (static_cast<int>(frameData.vertexBuffer.getRequirements().size) != imDrawData->TotalVtxCount)) {
```
You are comparing a number of vertex and a size of buffer!
And if the size required is lesser, re-use the buffer.

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+        }
+
+        // IndexBuffer
+        if ((static_cast<VkBuffer>(frameData.indexBuffer) == VK_NULL_HANDLE) || (static_cast<int>(frameData.indexBuffer.getRequirements().size) < imDrawData->TotalIdxCount)) {
```
You are comparing a number of indices and a size of buffer!
And if the size required is lesser, re-use the buffer.

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+
+                    VkResult result{VK_SUCCESS};
+                    if (!bufferBuilder.build(frameData.vertexBuffer, &result)) {
+                        LUG_LOG.error("Gui::initFontsTexture: Can't create staging buffer: {}", result);
```
We are not in `initFontsTexture`

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+
+                VkResult result{VK_SUCCESS};
+                if (!deviceMemoryBuilder.build(frameData.vertexMemory, &result)) {
+                    LUG_LOG.error("Gui::initFontsTexture: Can't create staging buffer device memory: {}", result);
```
We are not in `initFontsTexture`

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+
+                    VkResult result{VK_SUCCESS};
+                    if (!bufferBuilder.build(frameData.indexBuffer, &result)) {
+                        LUG_LOG.error("Gui::initFontsTexture: Can't create staging buffer: {}", result);
```
We are not in `initFontsTexture`

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+
+                VkResult result{VK_SUCCESS};
+                if (!deviceMemoryBuilder.build(frameData.indexMemory, &result)) {
+                    LUG_LOG.error("Gui::initFontsTexture: Can't create staging buffer device memory: {}", result);
```
We are not in `initFontsTexture`

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+
+        for (int n = 0; n < imDrawData->CmdListsCount; n++) {
+            const ImDrawList* cmd_list = imDrawData->CmdLists[n];
+            memcpy(vertexMemoryPtr, cmd_list->VtxBuffer.Data, cmd_list->VtxBuffer.Size * sizeof(ImDrawVert));
```
Can't you use `vertexBufferSize`?

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+        for (int n = 0; n < imDrawData->CmdListsCount; n++) {
+            const ImDrawList* cmd_list = imDrawData->CmdLists[n];
+            memcpy(vertexMemoryPtr, cmd_list->VtxBuffer.Data, cmd_list->VtxBuffer.Size * sizeof(ImDrawVert));
+            memcpy(indexMemoryPtr, cmd_list->IdxBuffer.Data, cmd_list->IdxBuffer.Size * sizeof(ImDrawIdx));
```
Can't you use `indexBufferSize`?

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+    FrameData& frameData = _framesData[currentImageIndex];
+
+    // Note: Alignment is done inside buffer creation
+    VkDeviceSize vertexBufferSize = imDrawData->TotalVtxCount * sizeof(ImDrawVert);
```
Use const

Quentin01 commented on src/lug/Graphics/Vulkan/Gui.cpp:
```diff
@@ -0,0 +1,876 @@
+
+    // Note: Alignment is done inside buffer creation
+    VkDeviceSize vertexBufferSize = imDrawData->TotalVtxCount * sizeof(ImDrawVert);
+    VkDeviceSize indexBufferSize = imDrawData->TotalIdxCount * sizeof(ImDrawIdx);
```
Use const

StuartSulaski committed:
Corrected all changes that were requested by Quentin01 except one

line 308 in Gui.cpp a function need to be written to replace
vkCmdCopyBufferToImage()

Quentin01 committed:
Fix compilation and minor thing
